### PR TITLE
MilliSecond Granualrity added to Copy operation on unix

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/corefx/issues/12403
without fix
```
Input: 187 ms, 636669266341879837 ticks
  Output: 0 ms, 504911232000000000 ticks
  Directory Created
  Input: 187 ms, 636669266341879837 ticks
  Output: 0 ms, 636669266340000000 ticks
```
with fix
```
  Input: 764 ms, 636669013077642750 ticks
  Output: 0 ms, 504910936800000000 ticks
  Directory Created
  Input: 764 ms, 636669013077642750 ticks
  Output: 764 ms, 636669013077642750 ticks
  Finished:    System.IO.FileSystem.Tests
```
The CI should fail on 4f23720 as fix is not there and pass on the recent commit

### CODE REVIEW

File: src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs [hunk 1]
```diff
@@ -64,6 +64,26 @@ public override IEnumerable<TimeFunction> TimeFunctions(bool requiresRoundtrippi
                 ((testFile, time) => { testFile.LastWriteTimeUtc = time; }),
                 ((testFile) => testFile.LastWriteTimeUtc),
                 DateTimeKind.Utc);
+            yield return TimeFunction.Create(
+                ((testFile, time) => CopytoOperation()),
+                ((testFile) => testFile.LastWriteTimeUtc),
+                DateTimeKind.Utc);
+        }
+
+        private void CopytoOperation()
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs]: Nit: to => To

File: src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs [hunk 2]
```diff
@@ -64,6 +64,26 @@ public override IEnumerable<TimeFunction> TimeFunctions(bool requiresRoundtrippi
                 ((testFile, time) => { testFile.LastWriteTimeUtc = time; }),
                 ((testFile) => testFile.LastWriteTimeUtc),
                 DateTimeKind.Utc);
+            yield return TimeFunction.Create(
+                ((testFile, time) => CopytoOperation()),
+                ((testFile) => testFile.LastWriteTimeUtc),
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs]: I don't understand how this fits in with this TimeFunction.Create.  The purpose of that helper as I understand it is to modify testFile in the first delegate and then validate with the second that it was appropriately changed.  But in this new case, testFile isn't being modified in the original delegate, and thus the second delegate is irrelevant.  Why is this new test then plugging into this in this way?

File: src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs [hunk 3]
```diff
@@ -64,6 +64,26 @@ public override IEnumerable<TimeFunction> TimeFunctions(bool requiresRoundtrippi
...
+        }
+
+        private void CopytoOperation()
+        {
+            string fileName = GetTestFileName();
+            FileInfo input = new FileInfo(Path.Combine(TestDirectory, fileName));
+            FileInfo output = new FileInfo(Path.Combine(TestDirectory, GetTestFileName(), fileName));
+            input.Create().Dispose();
+
+            Assert.NotEqual(0, input.LastWriteTime.Millisecond);
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs]: Won't this fail 1/1000 times?

File: src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs [hunk 4]
```diff
@@ -64,6 +64,26 @@ public override IEnumerable<TimeFunction> TimeFunctions(bool requiresRoundtrippi
...
+            Assert.Equal(0, output.LastWriteTime.Millisecond);
+            output.Directory.Create();
+            output = input.CopyTo(output.FullName, true);
+
+            Assert.NotEqual(0, input.LastWriteTime.Millisecond);
+            Assert.NotEqual(0, output.LastWriteTime.Millisecond);
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs]: Same here?

File: src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs (hunk 2 above)
Comment (Anipik) [reply src/System.IO.FileSystem/tests/FileInfo/GetSetTimes.cs]: I wanted to use the same behavior if we wanted to add some other similar tests in the future(Checking that the file still has millisecond attribute after some operation).

But i was not able to properly utilize existing this. For time being, i will just move this to its own test.

**Comment by Anipik** [comment]
@danmosemsft can you please review this one ?

File: src/Native/Unix/System.Native/pal_io.c [hunk 5]
```diff
@@ -1278,18 +1278,18 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
 #if HAVE_FUTIMES
         struct timeval origTimes[2];
         origTimes[0].tv_sec = sourceStat.st_atime;
-        origTimes[0].tv_usec = 0;
+        origTimes[0].tv_usec = ST_ATIME_NSEC(&sourceStat)/1000;
```
Comment (danmosemsft) [inline src/Native/Unix/System.Native/pal_io.c]: Nit, I believe we have single space around operators

**Pull request merged**